src/main/java/org/truffleruby/parser/lexer/RubyLexer.java

```diff
@@ -1798,7 +1798,7 @@ private int identifier(int c, boolean commandState) {
             }
             tempVal = createTokenRope();
 
-            if (result == 0 && Character.isUpperCase(tempVal.get(0) & 0xFF)) {
+            if (result == 0 && Character.isUpperCase(tempVal.getString().charAt(0))) {
```

Comment:
`charAt()` might not work if the character would be a surrogate and need 2 `char` to be represented.
`codePointAt(0)` is a possible fix for that.

I think it is possible to query directly via JCodings without converting to java.lang.String by using `Encoding#mbcToCode` and `Encoding#isUpper` (seems worth a helper method). That should be more efficient.